### PR TITLE
features/survey-refactor

### PR DESCRIPTION
**Pull request by joshcampbell191**
- Refactored PA survey
- Refactored Checkbox Widget
- Refactored server side file upload logic
- Added SurveyBase class
  - Added generic Error Summary
- Added File Meter Widget (file space consumed by uploads)
- Removed unused properties from PA survey json

### CODE REVIEW

File: ComplaintFormCore/ClientApp/src/survey.ts [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        }
+
+        this.displayErrorSummary(questionErrors);
+    }
+
+    // TODO: This method should actually be converted into a widget.
+    private registerCustomProperties(): void {
+        JsonObject.metaData.addProperty("itemvalue", {
+            name: "htmlAdditionalInfo:text"
+        });
```
Comment (joshcampbell191) [inline ComplaintFormCore/ClientApp/src/survey.ts:166]: Move to checkbox widget

File: ComplaintFormCore/ClientApp/src/widgets/filemeterwidget.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+     *
+     * @param activatedBy Tells how your widget has been activated by: property, type or customType.
+     * property - It means that it will activated if a property of the existing question type is set to particular value, for example inputType = "date".
+     * type - You are changing the behaviour of entire question type. For example render radiogroup question differently, have a fancy radio buttons.
+     * customType - You are creating a new type, like in our example "textwithbutton".
+     */
+    activatedByChanged(activatedBy: string): void {
+        this.activatedBy = activatedBy;
+        JsonObject.metaData.removeProperty("file", "showMeter");
+        JsonObject.metaData.removeProperty("file", "totalSize");
```
Comment (pierluc-ssoucy) [inline ComplaintFormCore/ClientApp/src/widgets/filemeterwidget.ts:31]: I understand the purpose, but is this really required?

File: ComplaintFormCore/ClientApp/src/widgets/checkboxwidget.ts [hunk 3]
```diff
@@ -41,16 +41,16 @@ export class CheckboxWidget extends Widget {
      */
     activatedByChanged(activatedBy: string): void {
         this.activatedBy = activatedBy;
-        Survey.JsonObject.metaData.removeProperty("checkbox", "renderAs");
+        JsonObject.metaData.removeProperty("checkbox", "renderAs");
```
Comment (pierluc-ssoucy) [inline ComplaintFormCore/ClientApp/src/widgets/checkboxwidget.ts:44]: There is a few instances of this one. Just by reading it, I find it slightly a bit confusing to not have the JsonObject class call prefixed with Survey.

File: ComplaintFormCore/ClientApp/src/index.ts [hunk 4]
```diff
@@ -63,13 +63,26 @@ declare let Symbol;
         globalThis.showPreview = SurveyNavigation.showPreview;
         globalThis.completeSurvey = SurveyNavigation.completeSurvey;
 
-        globalThis.initPaSurvey = (lang, token) => {
+        globalThis.initPaSurvey = async (lang: "fr" | "en", token) => {
             const jsonUrl = "/sample-data/survey_pa_complaint.json";
 
             CheckboxWidget.init();
+            FileMeterWidget.init();
```
Comment (pierluc-ssoucy) [inline ComplaintFormCore/ClientApp/src/index.ts:70]: This feels a bit odd, it's not the end of the world, but looks like its something that should be initialized automatically when used. If not we don't move forward with changing this, I feel like a name like "register" or something along those lines might better describe what's happening.

**Review by pierluc-ssoucy — CHANGES_REQUESTED** [review]

File: ComplaintFormCore/ClientApp/src/index.ts (hunk 4 above)
Comment (joshcampbell191) [reply ComplaintFormCore/ClientApp/src/index.ts]: These calls have been moved to `registerWidgets`.

File: ComplaintFormCore/ClientApp/src/index.ts [hunk 5]
```diff
@@ -63,13 +63,26 @@ declare let Symbol;
...
+
+            paSurvey.onPaSurveyComplete.add((survey: Survey.SurveyModel, options: any) => {
+                const fileNumber = document.getElementById("sp_survey_file_number");
+                if (fileNumber) {
+                    fileNumber.innerText = options.referenceNumber;
+                }
+            });
+
+            await paSurvey.loadSurveyFromUrl(jsonUrl);
+            paSurvey.render();
```
[the remark this answers is not in the document]
Comment (joshcampbell191) [reply ComplaintFormCore/ClientApp/src/index.ts:77]: A variable has been added to the `completedHtml` that gets set once the response is saved to the server.

File: ComplaintFormCore/ClientApp/src/index.ts [hunk 6]
```diff
@@ -63,13 +63,26 @@ declare let Symbol;
...
 
-            const paSurvey = new PaSurvey();
-            paSurvey.init(jsonUrl, lang, token);
+            /* const paSurvey = new PaSurvey();
+            paSurvey.init(jsonUrl, lang, token);*/
```
[the remark this answers is not in the document]
Comment (joshcampbell191) [reply ComplaintFormCore/ClientApp/src/index.ts]: Removed.

File: ComplaintFormCore/ClientApp/src/pa/newPaSurvey.ts [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        });
+
+        this.survey.onUploadFiles.add((sender: SurveyModel, options: any) => {
+            this.handleOnUploadFiles(sender, options);
+        });
+
+        this.survey.onClearFiles.add((sender: SurveyModel, options: any) => {
+            this.handleOnClearFiles(sender, options);
+        });
+    }
```
[the remark this answers is not in the document]
Comment (joshcampbell191) [reply ComplaintFormCore/ClientApp/src/pa/newPaSurvey.ts]: An abstract method `registerEventHandlers` has been added to the `SurveyBase` class.

File: ComplaintFormCore/ClientApp/src/pa/newPaSurvey.ts [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+                return;
+            }
+
+            options.complete();
+        })();
+    }
+
+    private handleOnComplete(sender: SurveyModel, options: any): void {
+        void (async () => {
+            const completeUrl = `/api/PASurvey/Complete?complaintId=${this.authToken}`;
```
[the remark this answers is not in the document]
Comment (joshcampbell191) [reply ComplaintFormCore/ClientApp/src/pa/newPaSurvey.ts:74]: Can be done at a later time.

File: ComplaintFormCore/ClientApp/src/survey.ts [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+    SurveyModel
+} from "survey-vue";
+import { Converter } from "showdown";
+import Vue from "vue";
+
+export abstract class SurveyBase extends Survey {
+    private converter = new Converter({
+        simpleLineBreaks: true,
+        tasklists: true
+    });
```
[the remark this answers is not in the document]
Comment (joshcampbell191) [reply ComplaintFormCore/ClientApp/src/survey.ts]: The property has been changed to readonly.

File: ComplaintFormCore/ClientApp/src/survey.ts [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+        let index = 1;
+        const list = document.createElement("ul");
+        questionErrors.forEach((errors: SurveyError[], question: Question) => {
+            const title = document.createElement("div");
+            title.innerHTML = question.processedTitle;
+            for (const error of errors) {
+                const item = document.createElement("li");
+                const link = document.createElement("a");
+                link.href = `#${question.inputId}`;
+                link.innerText = `Error ${index++}: ${title.innerText} - ${error.getText()}`;
```
[the remark this answers is not in the document]
Comment (joshcampbell191) [reply ComplaintFormCore/ClientApp/src/survey.ts]: Localization has been added through `surveyLocalization`.

File: ComplaintFormCore/ClientApp/src/survey.ts [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+                item.appendChild(link);
+                list.appendChild(item);
+            }
+        });
+
+        const header = document.createElement("h2");
+        header.innerText = `The form could not be submitted because ${index - 1} error(s) were found.`;
```
[the remark this answers is not in the document]
Comment (joshcampbell191) [reply ComplaintFormCore/ClientApp/src/survey.ts]: Localization has been added through `surveyLocalization`.

File: ComplaintFormCore/ClientApp/src/survey.ts (hunk 1 above)
Comment (joshcampbell191) [reply ComplaintFormCore/ClientApp/src/survey.ts]: The property has been moved to the checkbox widget.

File: ComplaintFormCore/ClientApp/src/widgets/filemeterwidget.ts (hunk 2 above)
Comment (joshcampbell191) [reply ComplaintFormCore/ClientApp/src/widgets/filemeterwidget.ts]: These have been removed as I could only find them in a single widget. Worse case, we can always come back and re-add it.

File: ComplaintFormCore/ClientApp/src/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (joshcampbell191) [reply ComplaintFormCore/ClientApp/src/index.ts]: These have been moved to the `registerWidgets` method of `NewPaSurvey`.

**Review by pierluc-ssoucy — APPROVED** [review]

**Pull request merged**